File: tqi-desafio-poo-dio/src/main/java/augusto108/ces/tqidesafiopoodio/dominio/entidades/Bootcamp.java
package augusto108.ces.tqidesafiopoodio.dominio.entidades;

import jakarta.persistence.*;

import java.time.LocalDateTime;
import java.util.LinkedHashSet;
import java.util.Set;

@Entity
@Table(name = "bootcamp")
public final class Bootcamp extends EntidadeBase {
    @Column(name = "descricao", nullable = false, unique = true, length = 45)
    private String descricao;

    @Column(name = "detalhamento", nullable = false)
    private String detalhamento;

    @Column(name = "data_inicio")
    private LocalDateTime dataInicio;

    @Column(name = "data_encerramento")
    private LocalDateTime dataEncerramento;

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
            name = "bootcamp_atividades",
            joinColumns = @JoinColumn(name = "bootcamp_id"), inverseJoinColumns = @JoinColumn(name = "atividade_id")
    )
    private final Set<Atividade> atividades = new LinkedHashSet<>();

    public Bootcamp() {
    }

    public Bootcamp(String descricao, String detalhamento) {
        this.descricao = descricao;
        this.detalhamento = detalhamento;
        this.dataInicio = LocalDateTime.now();
        this.dataEncerramento = dataInicio.plusDays(45);
    }

    public String getDescricao() {
        return descricao;
    }

    public void setDescricao(String descricao) {
        this.descricao = descricao;
    }

    public String getDetalhamento() {
        return detalhamento;
    }

    public void setDetalhamento(String detalhamento) {
        this.detalhamento = detalhamento;
    }

    public LocalDateTime getDataInicio() {
        return dataInicio;
    }

    public LocalDateTime getDataEncerramento() {
        return dataEncerramento;
    }

    public void setDataInicio(LocalDateTime dataInicio) {
        this.dataInicio = dataInicio;
    }

    public void setDataEncerramento(LocalDateTime dataEncerramento) {
        this.dataEncerramento = dataEncerramento;
    }

    public Set<Atividade> getAtividades() {
        return atividades;
    }

    @Override
    public String toString() {
        return "Bootcamp " + descricao;
    }
}

File: tqi-desafio-poo-dio/src/main/kotlin/augusto108/ces/bootcamptracker/model/entities/Bootcamp.kt
package augusto108.ces.bootcamptracker.model.entities

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonPropertyOrder
import jakarta.persistence.*
import java.time.LocalDateTime

@Entity
@Table(name = "bootcamp")
@JsonPropertyOrder("id", "description", "startDate", "finishDate", "details")
class Bootcamp @JsonCreator(mode = JsonCreator.Mode.PROPERTIES) constructor(
    @Column(name = "bootcamp_description", nullable = false, unique = true, length = 30) var description: String = "",
    @Column(name = "bootcamp_details", nullable = false, length = 80) var details: String = "",
    @Column(name = "start_date") var startDate: LocalDateTime? = null,
    @Column(name = "finish_date") var finishDate: LocalDateTime? = null,
    id: Int = 0
) : BaseEntity(id) {

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "bootcamp_activity",
        joinColumns = [JoinColumn(name = "bootcamp_id")],
        inverseJoinColumns = [JoinColumn(name = "activity_id")]
    )
    var activities: MutableSet<Activity> = HashSet()

    override fun toString(): String = description
}